File: acmicpc/acmicpc07576.py
import sys
from collections import deque

def bfs(n, m, box):
    dis = [-1, 1, 0, 0]
    djs = [0, 0, -1, 1]

    days = -1
    while queue:
        days += 1
        
        for _ in range(len(queue)):
            i, j = queue.popleft()
            for k in range(4):
                ni = i + dis[k]
                nj = j + djs[k]

                if (0 <= ni < n) and (0 <= nj < m) and (box[ni][nj] == 0):
                    box[ni][nj] = 1
                    queue.append([ni, nj])
    for lst in box:
        if 0 in lst:
            return -1

    return days


m, n = map(int, sys.stdin.readline().split())
box = []
queue = deque()

for i in range(n):
    row = list(map(int, sys.stdin.readline().split()))
    box.append(row)
    for j in range(m):
        if row[j] == 1:
            queue.append([i, j])

print(bfs(n, m, box))

File: acmicpc/acmicpc07576.cpp
#include <stdio.h>
#include <stdlib.h>

class Queue{
private:
    int right;
    int left;
    int arr[1000000][2];
public:
    Queue(){
        right = 0;
        left = 0;
    }
    void push(int a, int b){
        arr[right][0] = a;
        arr[right][1] = b;
        right += 1;
    }
    int* pop(){
        static int temp[2];
        temp[0] = arr[left][0];
        temp[1] = arr[left][1];
        left += 1;

        return temp;
    }
    int getsize(){
        return right - left;
    }
};

int main(){
    Queue queue;
    int m, n;
    scanf("%d", &m);
    scanf("%d", &n);
    
    
    int *box[1001];
    for (int i=0; i < n; i++){
        box[i] = (int *)malloc(sizeof(int) * 1001);
        for (int j=0; j < m; j++){
            scanf("%d", &box[i][j]);
            if (box[i][j] == 1){
                queue.push(i, j);
            } 
        }
    }

    int dis[4] = {-1, 1, 0, 0};
    int djs[4] = {0, 0, -1, 1};
    int days = -1;

    int* now;
    int ni, nj, len_queue;
    while (queue.getsize()){
        days += 1;

        len_queue = queue.getsize();
        for (int k=0; k < len_queue; k++){
            now = queue.pop();
            for (int i=0; i < 4; i++){
                ni = now[0] + dis[i];
                nj = now[1] + djs[i];

                if ((ni >= 0) && (ni < n) && (nj >= 0) && (nj < m) && (box[ni][nj] == 0)){
                    box[ni][nj] = 1;
                    queue.push(ni, nj);
                }
            }
        }
    }

    for (int i=0; i < n; i++){
        for (int j=0; j < m; j++){
            if (box[i][j] == 0){
                printf("%d\n", -1);
                return 0;
            }
        }
    }
    printf("%d\n", days);
    return 0;
}